db/versions/0091-migration.sql

```diff
@@ -0,0 +1,166 @@
+    visible,
+    pop_receipt
+  FROM azure_queue_messages
+  WHERE queue_name = 'claim-queue'
+  AND expires > now();
+
+  CREATE INDEX queue_claimed_task_idx ON queue_claimed_tasks (task_id, run_id);
+  CREATE INDEX queue_claimed_task_vis_idx ON queue_claimed_tasks (visible);
+  CREATE INDEX queue_claimed_task_queue_idx ON queue_claimed_tasks (task_queue_id, worker_group, worker_id);
+
+
+  -- Pending (scheduled)
+  -- purpose: keep the record of all tasks that were scheduled and claimed
+  -- queries:
+  --  * what tasks are pending for a given queue
+  --  * what tasks are claimed and still running (for claim expire purpose)
+  --  * find tasks that were claimed but not resolved (for claim expire purpose)
+  CREATE TABLE queue_pending_tasks (
+    task_queue_id text not null,
+    priority int not null,
+    task_id text not null, -- should be unique and a key
+    run_id integer not null,
+    hint_id text not null, -- to know that task was claimed properly, stored inside tasks.runs[]
+    inserted timestamptz not null,
+    expires timestamptz not null,
+    visible timestamptz not null,
+    pop_receipt uuid null
+  );
+
+  -- rest goes into queue_pending_tasks
+  INSERT INTO queue_pending_tasks
+    (task_queue_id, priority, task_id, run_id, hint_id, inserted, expires, visible, pop_receipt)
+  SELECT
+    task_queue_id,
+    priority,
+    convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'taskId',
+    CAST(convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'runId' AS INTEGER),
+    convert_from(decode(message_text, 'base64'), 'utf-8')::jsonb->>'hintId',
+    inserted,
+    expires,
+    visible,
+    pop_receipt
+  FROM azure_queue_messages
+  WHERE queue_name NOT IN ('claim-queue', 'deadline-queue', 'resolved-queue')
+  AND task_queue_id IS NOT NULL
+  AND expires > now();
+
+  CREATE UNIQUE INDEX queue_pending_task_idx ON queue_pending_tasks (task_id);
+  CREATE INDEX queue_pending_task_vis_idx ON queue_pending_tasks (visible, expires);
+  CREATE INDEX queue_pending_task_queue_idx ON queue_pending_tasks (task_queue_id, priority, inserted);
+
+  GRANT select, insert, update, delete ON queue_pending_tasks to $db_user_prefix$_queue;
+  GRANT select, insert, update, delete ON queue_task_deadlines to $db_user_prefix$_queue;
+  GRANT select, insert, update, delete ON queue_resolved_tasks to $db_user_prefix$_queue;
+  GRANT select, insert, update, delete ON queue_claimed_tasks to $db_user_prefix$_queue;
+
+  -- delete old data
+  -- or delete whole table? will there be workers still running at the time of migration?
+  -- maybe it's better to leave the table as is to let remaining workers finish what was started
+  -- and in worst case new workers might restart few tasks
+  -- other queues should not suffer much, as deadline/resolved/claims are handling errors
+  -- table can be dropped in a followup migration on next release
+  -- DELETE FROM azure_queue_messages;
```

Comment:
The migration should update the stored functions in the same transaction, so that once the transaction commits, nothing references the old table. So, it should be safe to delete it.

Reply:
So with this migration we are only marking old `azure_*` methods as deprecated. And we didn't drop functions ad deprecation time before, as far as I know. 

But since we also remove so much, maybe we should also drop existing methods?

Reply:
In this context, "deprecated" means "still exists and works as documented, but new code should not use it". As written, though, these functions will _not_ work as documented, since they address the old `azure_queue_messages` table.

Typically that's meant re-implementing the deprecated functions in terms of the new tables or columns. For example, when we changed from provisionerId/workerType to workerPoolId, the functions that referred to `provisioner_id` and `worker_type` were re-implemented to read the `worker_pool_id` column from the DB, split on `/`, and return the two parts as `provisioner_id` and `worker_type`. These deprecated functions only run during the time when _old_ code is running against a migrated DB.

Generally after a migration completes, the code is upgraded "soon" after, so the functions are only used for the duration of "soon". Maybe it's acceptable to have those deprecated functions just crash, to avoid the effort of re-implementing them in terms of the new tables. Maybe it makes sense for _some_ functions to just crash (e.g., those used by deadline resolver, which can crash-loop for a few minutes without any user-visible impact), while re-implementing others (e.g., putting messages in the pending queues, so that `queue.createTask` calls don't fail).

Reply:
Thanks Dustin! 
Totally makes sense. Those functions would be useless during migration and would lead to data loss likely.
I will try to repurpose those for new tables if it would be possible.